pwiz_tools/Skyline/Controls/Graphs/GraphFullScan.cs

```diff
@@ -412,10 +446,108 @@ private void CreateIonMobilityHeatmap()
+            var charges = stateProvider.ShowIonCharges(adducts);
+            var rankTypes = group.IsProteomic
+                ? settings.TransitionSettings.Filter.PeptideIonTypes
+                : settings.TransitionSettings.Filter.SmallMoleculeIonTypes;
+            var rankAdducts = group.IsProteomic
+                ? settings.TransitionSettings.Filter.PeptideProductCharges
+                : settings.TransitionSettings.Filter.SmallMoleculeFragmentAdducts;
+            var rankCharges = Adduct.OrderedAbsoluteChargeValues(rankAdducts);
+
+            int i = 0;
+            foreach (IonType type in rankTypes)
+            {
+                if (types.Remove(type))
+                    types.Insert(i++, type);
+            }
+
+            i = 0;
+            var showAdducts = new List<Adduct>();
+            foreach (var charge in rankCharges)
+            {
+                if (charges.Remove(charge))
+                    charges.Insert(i++, charge);
+                // NB for all adducts we just look at abs value of charge
+                // CONSIDER(bspratt): we may want finer per-adduct control for small molecule use
+                showAdducts.AddRange(adducts.Where(a => charge == Math.Abs(a.AdductCharge)));
+            }
+
+            showAdducts.AddRange(adducts.Where(a =>
+                charges.Contains(Math.Abs(a.AdductCharge)) && !showAdducts.Contains(a)));
+
+            var precursorNodePath = DocNodePath.GetNodePath(precursor.Id, _documentContainer.DocumentUI);
+            if (precursorNodePath != null) 
+            {
+                var spectrumInfo = new SpectrumPeaksInfo(Enumerable.Range(0, mzs.Count).Select(j => new SpectrumPeaksInfo.MI()
+                    {Mz = mzs[j], Intensity = (float)intensities[j]}).ToArray());
+
+                if (_rmis == null)
+                {
+                    _rmis = LibraryRankedSpectrumInfo.NewLibraryRankedSpectrumInfo(spectrumInfo,
+                        precursor.LabelType,
+                        precursor,
+                        settings,
+                        precursorNodePath.Peptide.SourceUnmodifiedTarget,
+                        precursorNodePath.Peptide.SourceExplicitMods,
+                        showAdducts,
+                        types,
+                        rankAdducts,
+                        rankTypes,
+                        null);
+
+                    //create an map of transition indices vs. peak rank to use in the mouse click and mouse move handlers.
+                    _transitionIndex = Enumerable.Repeat(-1, Enumerable.Max(_rmis.PeaksRanked.Select(p => p.Rank)) + 1).ToArray();
+                    foreach (var rankedPeak in _rmis.PeaksRanked)
+                    {
+                        try
+                        {
+                            var transitions = _msDataFileScanHelper.ScanProvider.Transitions;
+                            var tIndex = Enumerable.Range(0, transitions.Length).First(j =>

```

Comment:
I would recommend writing this as a regular "for" loop with a "break" in it.

Reply:
Rewrote as suggested